client/components/site-icon/index.jsx

```diff
@@ -70,6 +75,19 @@ const SiteIcon = React.createClass( {
 	}
 } );
 
-export default connect( ( state, { siteId } ) => (
-	siteId ? { site: getSite( state, siteId ) } : {}
-) )( SiteIcon );
+export default connect( ( state, { site, siteId, imgSize } ) => {
+	// Until sites state is completely within Redux, we provide compatibility
+	// in cases where site object is passed to use the icon.img property as URL
+	if ( site ) {
+		return {
+			iconUrl: get( site, 'icon.img' )
+		};
+	}
+
+	// Otherwise, assume we want to perform the lookup in Redux state
+	// exclusively using the site ID
+	return {
+		site: getSite( state, siteId ),
+		iconUrl: getSiteIconUrl( state, siteId, imgSize )
```

Comment:
>Why is this?

Mainly because otherwise every time I open my site selector, it would issue about 30 requests for each site shown, even if we already have data for those sites from `/me/sites`. We have this implied assumption in Calypso that sites data is expected to be universally requested via `/me/sites`, not specifically through any one query component. Not that this is very good, mind you, but the current situation nonetheless.